lib/web_ui/lib/ui_web/src/ui_web/initialization.dart

```diff
@@ -55,3 +55,33 @@ Future<void> bootstrapEngine({
     loader.didCreateEngineInitializer(bootstrap.prepareEngineInitializer());
   }
 }
+
+/// Switches the web engine into the web crawler mode.
+///
+/// In this mode the engine renders the semantics DOM tree in a way that's
+/// friendlier to web crawlers. For example, normally text appears only in
+/// `aria-label` attributes, but most crawlers ignore this attributes. When the
+/// crawler mode is enabled, the engine will put the text into an element as a
+/// DOM `Text` node (https://developer.mozilla.org/en-US/docs/Web/API/Text).
+///
+/// This function should only be called once. Calling it more than once will
+/// result in a `StateError` in debug mode, and it will have no effect in
+/// release mode.
+///
+/// It is expected that this function is called early in the app's
+/// initialization, before the framework renders the semantics tree, e.g. prior
+/// to invoking `runApp`. Calling it too late, such as after the first frame has
+/// been rendered or in response to a user action, may result in incorrect
+/// rendering of the semantics tree. The crawler may already have crawled the
+/// app and failed to find the data it needed to index it.
+void enableCrawlerMode() {
+  assert(
+    !_isCrawlerModeEnabled,
+    '`enableCrawlerMode` was called more than once.',
+  );
+  _isCrawlerModeEnabled = true;
+}
+
+/// Whether the crawler mode is enabled.
+bool get isCrawlerModeEnabled => _isCrawlerModeEnabled;
+bool _isCrawlerModeEnabled = false;
```

Comment:
Initializing `_isCrawlerModeEnabled` in the getter seems wrong. It means that you cannot call `enableCrawlerMode()` after calling the getter, i.e. the getter has a side-effect on the initialization lifecycle. For example, the following would throw an error with an incorrect error message:

```
if (!isCrawlerModeEnabled) {
  enableCrawlerMode(); // throws '`enableCrawlerMode` was called more than once.', which is wrong; it's called once
}
```

A better version of the getter might be:

```
bool get isCrawlerModeEnabled => isCrawlerModeEnabled ?? false;
```

However, now there's no difference between a nullable version and a non-nullable version. You either go from `null` to `true`, or from `false` to `true`.

Reply:
What I'm aiming at is making sure that after the engine reads the value of `isCrawlModeEnabled`, it shouldn't be overridable anymore. Doing it in the getter is easier, but I agree it's not ideal to have side effects in a getter.

Another way of doing it would be to have a 2nd `bool isCrawlModeSet` and a `preventFurtherCrawlModeChanges()`. We probably don't want to expose these through `dart:ui_web`. So you may have to move the `_isCrawlModeEnabled` into the private part of the engine, and expose it through `dart:ui_web` via a getter.

Given the extra complexity it would take to add this check, I'm not sure anymore if it's worth doing it.